### ISSUE TITLE
Is it possible to encode/decode non-primitive types inside Mapper?

### ISSUE THREAD

**Issue by jeffreyfjohnson** [Kotlin/kotlinx.serialization#404]
A project I am working on uses Google's Firestore NoSQL database, and the Android SDK takes a `Map<String, Any>` to create or update a record in the database.

The problem I'm working in is how to translate something like this:
```
data class Foo(val string: String, val number: Int, val timestamp: Instant)
```
into a map like this:
```
{
    "string" to "whatever string value",
    "number" to 1234,
    "timestamp" to [corresponding instance of java.util.Date]  // java.time.Instant instance translated to java.util.Date instance
}
```
Using encoders/decoders like the ones described [here](https://github.com/Kotlin/kotlinx.serialization/blob/master/example-jvm/src/NamedMapDemo.kt) allow me to go straight from a data class to a map, which is great.

My problem is that I don't know how to translate `Instant` into a non-primitive. I can obviously write a custom Serializer for Instant like so
```
@Serializer(forClass = Instant::class)
object InstantSerializer : KSerializer<Instant> {
        override val descriptor: SerialDescriptor =
            StringDescriptor.withName("Instant")

        override fun serialize(output: Encoder, obj: Instant) {
            output.encodeString(obj.toString())
        }

        override fun deserialize(input: Decoder): Instant {
            return Instant.parse(input.decodeString())
}
    
```

but then the map looks like

```
{
    "string" to "whatever string value",
    "number" to 1234,
    "timestamp" to "2019-03-11T00:00:00Z"
}
```

Which would be great for sending a HTTP request or something along those lines. The problem is that if the Firestore SDK picks up a `java.util.Date` in the `Map<String,Any>` that you pass it when creating/updating an object, it will store it differently than if it is a `String` representation of that time.

To sum it up, I'm wondering if it is possible to encode to/decode from an arbitrary, non-primitive type?

PS - I realize that timestamps in the ISO8601 format most likely can be sorted and queried against just as efficiently as any custom Date object that Firestore is using on their end, this is just the simplest example. There are others that would not work quite as well

**Comment by sandwwraith** [comment]
I'm not sure I've understood your question completely, but there is an example of writing a custom serializer called 'Representing class as multiple values': https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/custom_serializers.md#representing-classes-as-multiple-values

`Mapper` from runtime library should also be able to map/unmap nested structures, but it will flatten them. Possibly you want to just write another map in `encodeSerializableValue`...?

You can also try to convert everything to `JsonObject` first – it is smth like `Map<String, Any>` but Any is limited to other maps, lists or strings.

**Comment by jeffreyfjohnson** [comment]
Yeah, I was worried the question wasn't completely clear, since it is sort of an odd use case.

I don't necessarily need nested structures, or to represent a field as multiple values.

At it's most basic, I want to take a `data class`, and use serialization to output a **instance of `Date`** to a map. I don't want this output to be a `String` representation of a `Date` (e.g. `2019-03-11T00:00:00Z`), nor do I want it to be a `Long` representation (e.g. `1552921244`).

In order for it to work the way I'm hoping, the map that is output must contain a literal **instance of `Date`**.

The reason this is important for this use case is that the SDK that I am passing the `Map` into treats `Date`s differently

I hope this is more clear, but if not, I am happy to clarify in whatever way I can! Thank you for your help

**Comment by fvasco** [comment]
@jeffreyfjohnson are you looking for a serializer or for a bean mapper (ie Orika).

**Comment by sandwwraith** [comment]
Okay, I understand now. Well, this is a tough one... You can start with a default implementation of `Mapper` or `NamedMapDemo` and override `encodeSerializableValue` in such way that whenever it encounters some special descriptor/serializer (or a date instance) it writes it into a map 'as it is', otherwise calls default implementation.

**Comment by jeffreyfjohnson** [comment]
@fvasco I think we are looking for a serializer for this project

@sandwwraith That's an interesting idea. Is there any easy way to get the `tag` or the name of the field being written inside the encoder?

```
class MapEncoder(val map: MutableMap<String, Any> = mutableMapOf()) : NamedValueEncoder() {

    override fun beginCollection(
        desc: SerialDescriptor,
        collectionSize: Int,
        vararg typeParams: KSerializer<*>
    ): CompositeEncoder {
        encodeTaggedInt(nested("size"), collectionSize)
        return this
    }

    override fun encodeTaggedValue(tag: String, value: Any) {
        map[tag] = value
    }

    override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
        when {
            serializer is InstantSerializer && value is Instant -> {
                map.put("???", value)
            }
            else -> super.encodeSerializableValue(serializer, value)
        }
    }
}
```

In the code above, `encodeTaggedValue` is passed a `tag` that we can use as the key in the map. We don't get that passed in `encodeSerializableValue`. Is there a way to get that?

**Comment by sandwwraith** [comment]
I think `currentTag` property is what you're looking into.

**Comment by jeffreyfjohnson** [comment]
Great, thank you!

**Comment by jeffreyfjohnson** [comment]
In case anyone else stumbles upon this, this is my final encoder class:
```
class MapEncoder(val map: MutableMap<String, Any> = mutableMapOf()) : NamedValueEncoder() {

    override fun beginCollection(
        desc: SerialDescriptor,
        collectionSize: Int,
        vararg typeParams: KSerializer<*>
    ): CompositeEncoder {
        encodeTaggedInt(nested("size"), collectionSize)
        return this
    }

    override fun encodeTaggedValue(tag: String, value: Any) {
        map[tag] = value
    }

    override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
        when {
            serializer is InstantSerializer && value is Instant -> {
                map.put(currentTag, value)
                endStructure(serializer.descriptor) // need this, otherwise tag/key mapping gets messed up
            }
            else -> super.encodeSerializableValue(serializer, value)
        }
    }
}
```
